### PR TITLE
[DOC Beta] add API docs for component.js

### PR DESCRIPTION
**Pull request by toddjordan**
part of #14166 

moving documentation back into component.js

Couple of things of note: 
- lifecycle rule functions no longer defined on the Component prototype, but we still will call them if a child component implements them, so I went ahead and kept the documentation for them.  Lmk if I don't need to.  
- new component has more mixins.  I added these to `@uses` and made visible, this allows it to process inherited functions.
- there are a few new/changed public methods: `nearestChildOf` and `nearestInstanceOf` gone, but there now is a `nearestTypeOf`.  Also `isVisible` and `instrumentDisplay` gone as properties.

### CODE REVIEW

**Comment by toddjordan** [comment]
cc @rwjblue

File: packages/ember-glimmer/lib/component.js [hunk 1]
```diff
@@ -104,9 +199,282 @@ const Component = CoreView.extend(
...
+      {{yield}}
+      {{#if (hasBlock "inverse")}}
+        {{yield to="inverse"}}
+      {{else}}
+        How are you?
+      {{/if}}
+
+      ```
+      @public
+      @property hasBlock
```
Comment (rwjblue) [inline packages/ember-glimmer/lib/component.js]: We should kill this from the `Ember.Component` documentation, it is very confusing (as folks have reported before).  We need to document `hasBlock` / `has-block` like it was a template helper (like `{{each`, `{{with`, etc).

File: packages/ember-glimmer/lib/component.js [hunk 2]
```diff
@@ -24,6 +24,101 @@ export const IS_DISPATCHING_ATTRS = symbol('IS_DISPATCHING_ATTRS');
 export const HAS_BLOCK = symbol('HAS_BLOCK');
 export const BOUNDS = symbol('BOUNDS');
 
+
+/**
+@module ember
+@submodule ember-views
```
Comment (rwjblue) [inline packages/ember-glimmer/lib/component.js]: Lets change this to `ember-glimmer`

File: packages/ember-glimmer/lib/component.js [hunk 3]
```diff
@@ -104,9 +199,282 @@ const Component = CoreView.extend(
...
+      {{/foo-bar}}
+      {{! templates/components/foo-bar.hbs }}
+      {{#if hasBlockParams}}
+        This will be printed because a block was provided
+        {{yield this}}
+      {{/if}}
+      ```
+
+      @public
+      @property hasBlockParams
```
Comment (rwjblue) [inline packages/ember-glimmer/lib/component.js]: Same as `hasBlock`

**Comment by toddjordan** [comment]
Should I remove the mixins that are private from `@uses`?  Upside is it would remove the private mixins from the api doc, downside it would remove some documented private methods in `Ember.Component`.

File: packages/ember-views/lib/mixins/action_support.js [hunk 4]
```diff
@@ -1,3 +1,7 @@
+/**
+ @module ember
+ @submodule ember-views
```
Comment (rwjblue) [inline packages/ember-views/lib/mixins/action_support.js]: ember-glimmer

**Comment by rwjblue** [comment]
> Should I remove the mixins that are private from @uses?

Their class level docs should be marked as `@private` so that they don't show up in the left hand sidebar on emberjs.com/api/.

**Comment by toddjordan** [comment]
well, I remember doing that feature and then removing it (b/c of the other classes that reference it).  It was too come back when we were more fine grained with it, but at the time when this drops the classes will show up...

**Comment by rwjblue** [comment]
Hmm. I am definitely not in favor of these mixins showing up in the sidebar.  Lets leave them off for now then...

File: packages/ember-glimmer/lib/component.js [hunk 5]
```diff
@@ -104,9 +200,282 @@ const Component = CoreView.extend(
...
+
+    The parameters can then be referred to by enumerating over the list:
+
+    ```hbs
+    {{#each attrs.names as |name|}}{{name}}{{/each}}
+    ```
+
+    @static
+    @public
+    @property positionalParams
```
Comment (toddjordan) [inline packages/ember-glimmer/lib/component.js]: Should we remove `positionalParams` for the same reason as hasBlock/hasBlockParams?

Comment (rwjblue) [inline packages/ember-glimmer/lib/component.js]: No, because `positionalParams` is something you have to use in the JS side.

**Comment by toddjordan** [comment]
ok, addressed comments.  Did you see the third bullet above about missing public methods?  Is this ok?

**Comment by rwjblue** [comment]
Yes, thank you.  I'm 👍 on this.

**Comment by rwjblue** [comment]
Can you move https://github.com/emberjs/ember.js/blob/6c8abeb991a2355ff25ee7abe493aa68c4ced9eb/packages/ember-views/lib/mixins/aria_role_support.js#L14-L27 here and delete that file?  Also, we do not need the property to be initialized to `null` like that, we just need the docs here...

**Comment by toddjordan** [comment]
done

**Comment by rwjblue** [comment]
Looks like `ember-glimmer/component` is still importing `ember-views/mixins/aria_role_support` (causing errors).

**Comment by toddjordan** [comment]
ah shoot.  fixing

**Comment by rwjblue** [comment]
Thank you!

**Pull request merged**